cmd/outline-ss-server/main.go

```diff
@@ -59,68 +60,71 @@ func init() {
 	logger = logging.MustGetLogger("")
 }
 
-type ssPort struct {
-	tcpListener *net.TCPListener
-	packetConn  net.PacketConn
-	cipherList  service.CipherList
+type ssListener struct {
+	io.Closer
+	cipherList service.CipherList
 }
 
 type SSServer struct {
 	natTimeout  time.Duration
 	m           *outlineMetrics
 	replayCache service.ReplayCache
-	ports       map[int]*ssPort
+	listeners   map[string]*ssListener
 }
 
-func (s *SSServer) startPort(portNum int) error {
-	listener, err := net.ListenTCP("tcp", &net.TCPAddr{Port: portNum})
-	if err != nil {
-		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks TCP service failed to start on port %v: %w", portNum, err)
+func (s *SSServer) serve(listener io.Closer, cipherList service.CipherList) error {
```

Comment:
io.Closer as a listener is confusing.

I wonder if we should actually define a `Listener` type.
Perhaps we switch on the connection type instead of the Listener type.

So the serve logic is shared, with the look and calls to `Listener.Accept()`. Then based on the connection type you can dispatch to the appropriate handler.

I'm not sure yet which one is better. I have the impression that treating all the listeners the same will make things simpler.